math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -848,6 +848,59 @@ RVec<RVec<typename RVec<T>::size_type>> Combinations(const RVec<T>& v, const typ
    }
 }
 
+/// Return the indices of the elements which are not zero
+template <typename T>
+RVec<typename RVec<T>::size_type> Nonzero(const RVec<T> &v)
+{
+   using size_type = typename RVec<T>::size_type;
+   RVec<size_type> r;
+   for(size_type i=0; i<v.size(); i++) {
+      if(v[i] != 0) {
+         r.emplace_back(i);
+      }
+   }
+   return r;
+}
+
+/// Return the intersection of elements of two vectors.
+/// Each element of v1 is looked up in v2 and added to the returned vector if
+/// found. Following, the order of v1 is reserved. Repetitions of elements in
+/// v2 are added multiple times to the returned vector if allow_repetitions is
+/// set to true;
+template <typename T>
+RVec<T> Intersect(const RVec<T>& v1, const RVec<T>& v2, bool allow_repetitions = false)
+{
+   // TODO: Remove comment after review.
+   // NOTE: This can be done much more efficient (for large vectors) if we sort
+   // v1 and v2 before.
+   RVec<T> r;
+   using size_type = typename RVec<T>::size_type;
+   for(size_type i=0; i<v1.size(); i++) {
+      for(size_type j=0; j<v2.size(); j++) {
+         if(v1[i] == v2[j]) {
+            r.emplace_back(v1[i]);
+            if(!allow_repetitions) {
```

Comment:
Why did you think to include this?  I guess this would arise only if there were repetitions in the input `v1` RVec, in which case I expect the user would enable this option.  If there aren't repetitions in `v1` then this option won't have any impact.  And then if the user does have repetitions in `v1` and doesn't want them there I think they should call a dedicated function to remove them.  What do you think?

Reply:
I agree that the `allow_repetitions` is not really useful. I would like to change it in the following way: Make clear in the description of the function, that we simply return all elements in `v1` that are as well present in `v2`. Imho, this is exactly what the function should do.

@benkrikler Do you agree?

Reply:
Btw, then, we can go back to the suggestion you've done above and sort `v2` (or not if a special flag is set in the signature of the function).

Reply:
Yes, this sounds like a good idea to me.  I've just noticed you actually put something similar in the comments above.  I suppose there will be a set of cases when the input order of `v1` is not important either, in which case first sorting `v1` might be good for speed too?

Reply:
More precisely, I'm imagining something like:
```c++
template <typename T>
RVec<T> Intersect(const RVec<T>& v1, const RVec<T>& v2, bool v2_is_ordered = false, bool sort_v1 = false){
  auto v2_sorted = v2_is_ordered ? v2 : std::sort(v2);
  RVect<T> v_out; // is it possible / worth guessing an output size?
  if (sort_v1) {
    auto v1_sorted = std::sort(v1.begin(), v1.end());  
    std::set_intersection(v1_sorted.begin(), v1_sorted.end(), v2_sorted.begin(), v2_sorted.end(), v_out.begin());
  } else {
    // current implementation goes here:
    for (auto ele: v1){
      // Since we're assuming v2 to be sorted, std::binary_search might be viable instead of the containers own find method
      if (v2.find(ele) != v2.end()) v_out.emplace_back(ele);
    }
  }
  return v_out;
  ```

Reply:
`set_intersection` does not work out if you want to contain the order of `v1`, which seems for me a very crucial feature. I've decided for the following strategy: Sort `v2` (or toggle it off with `v2_is_sorted=true`) and use the `std::binary_search` on `v2` for each element of `v1`. The resulting time complexity is `O(N*log(N))` compared to `O(N*N)` of the brute force version. However, if we would assume that `v1` is sorted (or we sort it), then we could run in `O(N)`...

Reply:
Adding the `sort_v1` option would solve this by leaving it up the user to say whether or not we can use `set_intersection` by first sorting `v1`.  That would allow a user who doesn't care about v1's order to get this running quicker, but keeps the default behaviour as we originally needed, maintaining the ordering of `v1`.  

A subtle alternative would be to just check that `v1` is sorted using `std::is_sorted`.  That would simplify the user interface, but means that we always run the check for a sorted list.  Since that check is itself O(N) I suppose this isn't too bad an addition (and it's better than sorting `v1` ourselves I guess).